examples/hamiltonian_representation.py

```diff
@@ -0,0 +1,243 @@
+        elif isinstance(boolean_expr, Or):
+            hamiltonian = HamiltonianPolynomial.O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Xor):
+            hamiltonian = HamiltonianPolynomial.O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - 2.0 * hamiltonian * sub_hamiltonian
+        return hamiltonian
+
+    raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def get_name_to_id(boolean_exprs: Sequence[Expr]) -> Dict[str, int]:
+    """Maps the variables to a unique integer.
+
+    Args:
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+
+    # For run-to-run identicalness, we sort the symbol name lexicographically.
+    symbol_names = sorted(
+        {symbol.name for boolean_expr in boolean_exprs for symbol in boolean_expr.free_symbols}
+    )
+    return {symbol_name: i for i, symbol_name in enumerate(symbol_names)}
+
+
+def _gray_code_comparator(k1, k2, flip=False):
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def build_circuit_from_hamiltonians(
+    hamiltonian_polynomial_list: List[HamiltonianPolynomial],
+    qubits: List[cirq.NamedQubit],
+    theta: float,
+) -> cirq.Circuit:
+    """Builds a circuit according to [1].
+
+    Args:
+        hamiltonian_polynomial_list: the list of Hamiltonians, typically built by calling
+            build_hamiltonian_from_boolean().
+        qubits: The list of qubits corresponding to the variables.
+        theta: A single float scaling the rotations.
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+    combined = sum(hamiltonian_polynomial_list, HamiltonianPolynomial.O())
+
+    circuit = cirq.Circuit()
+
+    # Here we follow improvements of [4] cancelling out the CNOTs. The first step is to order by
+    # Gray code so that as few as possible gates are changed.
+    sorted_hs = sorted(
+        list(combined.hamiltonians.keys()), key=functools.cmp_to_key(_gray_code_comparator)
+    )
+
+    def _apply_cnots(previous_h, h):
+        # This function applies in sequence the CNOTs from previous_h and then h. However, given
+        # that the h are sorted in Gray ordering and that some cancel each other, we can reduce the
+        # number of gates. See [4] for more details.
+
+        # We first test whether the rotation is applied on the same qubit.
+        last_qubit_is_same = previous_h and h and previous_h[-1] == h[-1]
```

Comment:
Do we really need this `last_qubit_is_same` condition?  
Note that to implement `exp(i\theta ZZZ)`, the following two forms are equivalent:
1. CNOT(3, 1) CNOT(2, 1) Rz_1(\theta) CNOT(2, 1) CNOT(3, 1)
2. CNOT(3, 2) CNOT(2, 1) Rz_1(\theta) CNOT(2, 1) CNOT(3, 2)

As long as there are common binary representation bits in `Z` string, you can cancel the CNOT gates.

Reply:
Sorry, I don't quite understand what you suggest I do. I do agree with the equalities that you mentioned, I think it's figure 3 of the paper.

I also do agree that we can do the Rz on any qubit that has an Z, but choosing a different one also means changing how the CNOTs are applied. Don't they say that they "_adopt the convention in 3(b) where the CNOTs are always targeted on the highest order qubit possible_"? I mean that yes, we could change the applied qubit, but at some point, we are going to have to change it.

However, I am missing something indeed. They equations 8, 9, 10, and 11 can be applied to simplify the circuit further, but I don't understand in which order to apply them. It seems there's a general procedure, as indicated by footnote 10, but I couldn't find a new paper.

I will continue reading the paper of course, but if you have a hint, that would be helpful. In the meantime, I had realized that some of the unit tests were not very thorough, so I added some more. In particular, I'm focusing on this one:

`
(x0 ^ x1 ^ x2) | (x2 ^ x3 ^ x4)

(x0 ^ x1 ^ x2) | (x2 ^ x3 ^ x4)

                               ┌──┐
x0: ───@─────────────────────@─────@──────────────────────────@─────────────────────────────────
       │                     │     │                          │
x1: ───┼───@─────────────────┼────@┼────@─────────────────────┼───@─────────────────────────────
       │   │                 │    ││    │                     │   │
x2: ───X───X───Rz(-0.025π)───X────X┼────┼─────────────────────┼───┼───@─────────────────@───────
                                   │    │                     │   │   │                 │
x3: ───────────────────────────────┼────┼───@─────────────────┼───┼───┼─────────────────┼───@───
                                   │    │   │                 │   │   │                 │   │
x4: ───────────────────────────────X────X───X───Rz(-0.025π)───X───X───X───Rz(-0.025π)───X───X───
`

Reply:
Sorry, I don't have a plan in detail on this but can elaborate more about what I want to express and why intuitively I think `last_qubit_is_same` can be removed.  Simple example:
We want to have a rotation on ZZZ and ZZI. In Hamiltonian, it is `{(0, 1, 2): w1}` and `{(0, 1): w2}`. They don't have the same last qubit for rotation (assuming we always using the largest number for rotation one). However, the CNOT between them still can be canceled. Look at the following two realization:
![image](https://user-images.githubusercontent.com/16711681/114960791-9de16780-9e1c-11eb-8bb5-563644f4b988.png)
Versus
![image](https://user-images.githubusercontent.com/16711681/114960804-a3d74880-9e1c-11eb-9730-7dce793a647b.png)
(BTW, i cannot read the ASCII plot you posted previously. Using SVG plot is better here.)
They are the equivalent circuit but the second case has two less CNOT gates. 

I also realize it may not be easy to have Programmable. The key should be the eq. 10,-14 as you pointed out. But I need more time to re-digest that in this case. Will add more thoughts if I have more concrete ideas.

Reply:
I modified the code and test. I have implemented some of the simplifications, but I still need to implement equation 11.

It seems like it's a greedy approach, though, so I am not sure it's optimal. Nonetheless, it is still a superset of what I had before.

I will continue to implement equation 11 unless you have objections.

Reply:
I implemented equation 11. I think the circuit you had above wasn't an example of such simplification.

The algo is greedy, but it could be optimal too. I haven't wrapped my head around it. Maybe the margin was too small for their truly remarkable proof, so they opted for a footnote :)

Either way, I think I addressed your suggestion, but happy to do more if not.